### PR TITLE
Add ktlint static analysis, code auto-formatting and AS code style

### PR DESCRIPTION
**Pull request by sebaslogen**
## Why is this important?
To have consistent code following a standard formatting of spaces, general code and Compose functions+parameters.
No bikeshedding discussions about spaces, order of parameters or camelCases: 

## Any review notes?
- Added library [ktlint-gradle](https://github.com/jlleitschuh/ktlint-gradle) that integrates ktlint in gradle tasks
- Added standard [rules for Compose code conventions](https://mrmans0n.github.io/compose-rules/)
- Updated Readme with instructions on how to use ktlint and how to add it to your git pre-commit hooks
- Added Android Studio code style configuration that aligns with ktlint rules
- Auto-formatted all project files and manually corrected following suggested standards

### CODE REVIEW

**Comment by Frank1234** [comment]
Did you use this for a few months already, with > 1 devs? Do you still like it?

I did, and didn't like it. We turned it off eventually. I'm also not sure what problem it solves, really. Maybe let's discuss it IRL again and try it out on one more project before merging.

@ninovanhooff do you use this at HEMA? Would you like to?

**Comment by Frank1234** [comment]
I remember one of the problems we had in the LY car code, was that we would write ugly code and press format-code all the time, like most devs do. However, all code would jump if you did that, because it was formatted with a different formatter (ktlint), different from AS. We had commit rules for ktlint, but that is late in the process, resulting in the above.

We tried to import the ktlint rules into the IDE, there are tools for that, but not all rules can be imported. Same problem persists then. We eventually removed ktlint.

**Comment by sebaslogen** [comment]
> I remember one of the problems we had in the LY car code, was that we would write ugly code and press format-code all the time, like most devs do. However, all code would jump if you did that, because it was formatted with a different formatter (ktlint), different from AS. We had commit rules for ktlint, but that is late in the process, resulting in the above.
> 
> We tried to import the ktlint rules into the IDE, there are tools for that, but not all rules can be imported. Same problem persists then. We eventually removed ktlint.

Thanks for the experience and clarification.
This PR includes the `.idea/codeStyles/codeStyleConfig.xml` and the instructions to hook the check at commit time. Finding the problems very soon after making them is crucial to have a good dev experience. I've had problems with linters in the past that only failed on CI, and that was also not acceptable for me.

One solution I used long time ago and it's also supported by this one (but out of the scope of this PR) to the problem of too much lint errors when integrating lint for the first time, is to create a baseline, so lint tool only warns about new errors and leaves current ones, until you touch the file again.

**Comment by ninovanhooff** [comment]
## My current setup
### Android:
I use the AS defaults which work really well at the moment. But with every new developer on the team, I have to stress that it is important to read the README and tick some boxes in settings so that the built-in style is applied on save/commit. It would be nice to not have to deal with that.

### Typescript
tslint is used. I experience this as a bore / annoying / overhead. But necessary because we use 2 IDE's.

I agree with Frank. It would like to see it being a success in multiple projects that have multiple developers that are all happy with it.

File: core/actionresult/src/main/kotlin/nl/q42/template/actionresult/data/ActionResultMapper.kt [hunk 1]
```diff
@@ -3,82 +3,98 @@ package nl.q42.template.actionresult.data
...
-    try {
-        request().networkResponseToActionResult()
-    } catch (e: Exception) {
-        if (e is CancellationException) ActionResult.Error.Cancelled(e) else ActionResult.Error.Other(e)
+suspend fun <T : Any> mapToActionResult(request: suspend () -> NetworkResponse<T, ApiErrorResponse>): ActionResult<T> = try {
+    request().networkResponseToActionResult()
+} catch (e: Exception) {
+    if (e is CancellationException) {
+        ActionResult.Error.Cancelled(
+            e
```
Comment (ninovanhooff) [inline core/actionresult/src/main/kotlin/nl/q42/template/actionresult/data/ActionResultMapper.kt:32]: Is this rule "arguments on newlines, always"? Not a great result here

File: core/actionresult/src/main/kotlin/nl/q42/template/actionresult/data/ActionResultMapper.kt [hunk 2]
```diff
@@ -3,82 +3,98 @@ package nl.q42.template.actionresult.data
...
+private fun <T : Any> NetworkResponse<T, ApiErrorResponse>.networkResponseToActionResult(): ActionResult<T> = when (this) {
+    is NetworkResponse.Success -> ActionResult.Success(this.body)
+    is NetworkResponse.ServerError -> {
+        val errorMessage: String? = this.body?.message
+        val exception = Exception("$errorMessage $this.code")
+        when {
+            isNotFound(this.code) -> ActionResult.Error.NotFoundError
+            isTooManyRequests(this.code) -> ActionResult.Error.TooManyRequests(exception)
+            isUnAuthorized(
+                this.code
```
Comment (ninovanhooff) [inline core/actionresult/src/main/kotlin/nl/q42/template/actionresult/data/ActionResultMapper.kt:48]: yeah, this is not great

File: core/actionresult/src/main/kotlin/nl/q42/template/actionresult/data/ActionResultMapper.kt (hunk 1 above)
Comment (ninovanhooff) [reply core/actionresult/src/main/kotlin/nl/q42/template/actionresult/data/ActionResultMapper.kt:32]: oh I think it's the line length limit

File: core/navigation/src/main/kotlin/nl/q42/template/navigation/di/NavigationModule.kt [hunk 3]
```diff
@@ -11,8 +11,7 @@ import nl.q42.template.navigation.viewmodel.RouteNavigator
 @Module
 @InstallIn(ViewModelComponent::class)
 class NavigationModule {
-
     @Provides
     @ViewModelScoped
     fun bindRouteNavigator(): RouteNavigator = MyRouteNavigator()
-}
\ No newline at end of file
+}
```
Comment (ninovanhooff) [inline core/navigation/src/main/kotlin/nl/q42/template/navigation/di/NavigationModule.kt:17]: Ah great, empty last line is enforced 👍

File: core/ui/src/main/kotlin/nl/q42/template/ui/compose/OnLifecycleEvent.kt [hunk 4]
```diff
@@ -27,12 +27,13 @@ fun OnLifecycleEvent(onEvent: (event: Lifecycle.Event) -> Unit) {
     val eventHandler = rememberUpdatedState(onEvent)
     val lifecycleOwner by rememberUpdatedState(LocalLifecycleOwner.current)
     DisposableEffect(lifecycleOwner) {
-        val observer = LifecycleEventObserver { _, event ->
```
Comment (ninovanhooff) [inline core/ui/src/main/kotlin/nl/q42/template/ui/compose/OnLifecycleEvent.kt:30]: really weird that this would be broken into multi-line. Only 52 chars?

**Comment by Frank1234** [comment]
So how about using this at HEMA for a month @ninovanhooff , and then see how much / if you three like it? 

We can merge this PR afterwards.

**Comment by ninovanhooff** [comment]
Sure, I'll add it to my todo

File: core/actionresult/src/main/kotlin/nl/q42/template/actionresult/data/ActionResultMapper.kt (hunk 1 above)
Comment (sebaslogen) [reply core/actionresult/src/main/kotlin/nl/q42/template/actionresult/data/ActionResultMapper.kt:32]: I don't think it's the max line length, but maybe it's a matter of playing with the settings to find the ones that go well with our preferences and set it as the standard for the team/Q42

File: core/actionresult/src/main/kotlin/nl/q42/template/actionresult/data/ActionResultMapper.kt (hunk 2 above)
Comment (sebaslogen) [reply core/actionresult/src/main/kotlin/nl/q42/template/actionresult/data/ActionResultMapper.kt:48]: max_line_length = 140 in [.editorconfig](https://github.com/Q42/Template.Android/pull/50/files#diff-0947e2727d6bad8cd0ac4122f5314bb5b04e337393075bc4b5ef143b17fcbd5b)

https://pinterest.github.io/ktlint/latest/rules/standard/#max-line-length

**Comment by sebaslogen** [comment]
No need to worry about it @ninovanhooff you didn't fail me.
The idea failed to work with you, it failed with this PR and I personally conclude that the idea of static analysis doesn't currently fit at Q. Probably because most Android repos only have 2 devs at the most committing code at the same time, so very little space for inconsistencies that need to be fixed with a linter.

**Pull request closed without merging**